src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs

```diff
@@ -112,30 +257,22 @@ private unsafe int ReadCore(Span<byte> buffer)
             Debug.Assert(_handle != null);
             DebugAssertHandleValid(_handle);
 
-            // For named pipes, receive on the socket.
-            Socket? socket = _handle.NamedPipeSocket;
-            if (socket != null)
+            if (buffer.Length == 0)
             {
-                // For a blocking socket, we could simply use the same Read syscall as is done
-                // for reading an anonymous pipe.  However, for a non-blocking socket, Read could
-                // end up returning EWOULDBLOCK rather than blocking waiting for data.  Such a case
-                // is already handled by Socket.Receive, so we use it here.
-                try
-                {
-                    return socket.Receive(buffer, SocketFlags.None);
-                }
-                catch (SocketException e)
-                {
-                    throw GetIOExceptionForSocketException(e);
-                }
+                return 0;
             }
```

Comment:
We were just having a separate conversation about how 0-byte reads shouldn't be short-circuited.  Do we need this check?  Any reason we can't let it progress to the socket and block waiting for data to arrive?

Reply:
I think `StreamConformanceTests` `BlocksOnZeroByteReads` verifies this remains backwards compatible.

I'll change it, and see if a test fails.

Reply:
> I think StreamConformanceTests BlocksOnZeroByteReads verifies this remains backwards compatible.

I think that's a bug.  I suggest we fix it.  For example, we had to special-case the FileStream-wrapping-an-anonymous-pipe tests to have BlocksOnZeroByteReads as true only on Windows:
https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/System.IO.FileSystem/tests/FileStream/FileStreamConformanceTests.cs#L80
It'd be great if we could smooth that all out as part of this.

Reply:
@stephentoub I took a shot at this. It's not just a matter of removing these short-circuits, and update `BlocksOnZeroByteReads`. The 0-byte `read` calls don't block. I prefer to move this out of scope for this PR.

Reply:
Ok, thanks for trying.  Let's aim to solve that next.